File: Models_stability_plasticity_dilemma/biorxiv2018/Matlab_code/RW_synaptic.m
%% Defining amount of loops
Rep=10;                     % amount of replications
Tr=240;                      % amount of trials
betas=11;                   % beta iterations
Beta=0:1/(betas-1):1;       %learning rate values

%% other variables
POT_1=Tr/3;                 %point of switch to task rule 2 (trial 20)
POT_2=2*Tr/3;               %switch again to task rule 1    (trial 40)
part1=1:POT_1;              %first part
part2=POT_1+1:POT_2;        %second part
part3=POT_2+1:Tr;           %third part
nUnits=4;                   %model units
threshold=10;              %response threshold

%Input patterns
Activation=zeros(nUnits,2);
Activation(:,1)=[1,0,0,0];
Activation(:,2)=[0,1,0,0];

%% learning objectives
objective=zeros(nUnits,nUnits,Tr); 
objective(1,3,part1)=1;
objective(2,4,part1)=1;
objective(2,3,part2)=1;     
objective(1,4,part2)=1;
objective(1,3,part3)=1;  
objective(2,4,part3)=1; 

%% simulation loops
for b=1:betas                     %gradual change of beta 
    for r=1:Rep             %replication loop
%% model build-up
%processing layer
Rate=zeros(nUnits,Tr);    %rate neurons (matrix definieren)

%weights
W=zeros(nUnits,nUnits,Tr);  %matrix met gewichten definieren
W(1:2,3:4,1)=rand(2,2);       %initial weigth strengths (in het begin zijn alle stimuli (rij 1 en 2) gelijk verbonden met alle responsen (kolom 3 en 4)

%% Input
%randomization of input patterns
In=repmat(1:2,3,(POT_1));
Input=zeros(1,Tr);
Input(1,part1)=In(1,randperm(POT_1));
Input(1,part2)=In(2,randperm(POT_1));
Input(1,part3)=In(3,randperm(POT_1));

%% Other
Z=zeros(nUnits,Tr);             %input matrix
response=zeros(nUnits/2,Tr);    %response record
rew=zeros(1,Tr);                %reward or accuracy

%% the model

    for trial=1:Tr          %trial loop
        
        Z(:,trial)=Activation(:,Input(1,trial));

        %updating rate code units
        Rate(1:2,trial)=Z(1:2,trial);   
        Rate(3:4,trial)=max(0,Rate(1:2,trial)'*squeeze(W(1:2,3:4,trial))); 
            
        %response determination:
        if Rate(3,trial)>Rate(4,trial)    %response 1
            response(1,trial)= 1;
            response(2,trial)= 0;    
        else
            response(1,trial)= 0;
            response(2,trial)= 1;
        end;
        
        %reward value determination
        if squeeze(objective(Input(1,trial),3:4,trial))'==response(:,trial)
            rew(1,trial)=1;
        else
            rew(1,trial)=0;
        end; 
        
        %RW learning
        for p=1:2
            for q=3:4
                %weight updating (only for weights different than zero)
                W(p,q,trial+1)=W(p,q,trial)+Beta(1,b)*(objective(p,q,trial)-Rate(q,trial))*Rate(p,trial);
            end;
        end;
    prog=trial
    end;
    save(['Beta',num2str(b),'Rep',num2str(r),'_RWonly']); %write data to file with beta iteration, epsilon iteration and replication as name
    end; 
end;

File: Models_stability_plasticity_dilemma/biorxiv2018/Python_code/RWsynaptic.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Mon Mar  5 13:57:31 2018

@author: pieter
"""
import numpy as np
import pickle

# Defining amount of loops
Rep=10                          # amount of replications
Tr=240                          # amount of trials
betas=11                        # beta iterations
Beta=np.arange(0,1.1,0.1)       #learning rate values

# other variables
POT_1=Tr//3                     #point of switch to task rule 2 (trial 20)
POT_2=2*Tr//3                   #switch again to task rule 1    (trial 40)
part1=np.arange(POT_1)          #first part
part2=np.arange(POT_1,POT_2)    #second part
part3=np.arange(POT_2,Tr)       #third part
nUnits=4                        #model units

#Input patterns
Activation=np.zeros((nUnits,2))
Activation[:,0]=np.array([1,0,0,0])
Activation[:,1]=np.array([0,1,0,0])

#learning objectives
objective=np.zeros((nUnits,nUnits,Tr)) 
objective[0,2,part1]=1
objective[1,3,part1]=1
objective[1,2,part2]=1    
objective[0,3,part2]=1 
objective[1,3,part3]=1
objective[0,2,part3]=1

# simulation loops
for b in range(betas):           #gradual change of beta 
    for r in range(Rep):         #replication loop

        # model build-up
        #processing layer
        Rate=np.zeros((nUnits,Tr))        #rate neurons (matrix definieren)

        #weights
        W=np.zeros((nUnits,nUnits,Tr+1))          #matrix met gewichten definieren
        W[0:2,2:4,0]=np.random.random((2,2))    #initial weigth strengths

        # Input
        #randomization of input patterns
        In=np.tile([0,1],(3,POT_1))
        Input=np.zeros((1,Tr))
        Input[0,part1]=In[0,np.random.permutation(POT_1)]
        Input[0,part2]=In[1,np.random.permutation(POT_1)]
        Input[0,part3]=In[2,np.random.permutation(POT_1)]

        # Other
        Z=np.zeros((nUnits,Tr))            #input matrix
        response=np.zeros((nUnits//2,Tr))  #response record
        rew=np.zeros((1,Tr));              #reward or accuracy

        # the model

        for trial in range(Tr):            #trial loop
        
            Z[:,trial]=Activation[:,int(Input[0,trial])] 
                
            #updating rate code units
            for units in range(nUnits):
                Rate[units,trial]=Z[units,trial]+ np.matmul(np.transpose(Rate[:,trial]),np.squeeze(W[:,units,trial]))
                    
                #response determination:
                if Rate[2,trial]>Rate[3,trial]:   #responseone
                    response[0,trial]= 1
                    response[1,trial]= 0    
                else:  #responsetwo
                    response[0,trial]= 0
                    response[1,trial]= 1
        
            #reward value determination
            if np.all(np.transpose(np.squeeze(objective[int(Input[0,trial]),range(2,4),trial]))==response[:,trial]):
                rew[0,trial]=1
            else:
                rew[0,trial]=0 
        
            #RW learning
            for p in range(2):
                for q in range(2,4):
                    #weight updating (only for weights different than zero)
                    W[p,q,trial+1]=W[p,q,trial]+Beta[b]*(objective[p,q,trial]-Rate[q,trial])*Rate[p,trial]
        
        data_pickle = "Beta"+str(Beta[b])+"_Rep"+str(r)+"_RWonly.p"

        variables_str = ["rew","W"]
        variables =     [rew,W]
        data = []
        for loop in range(len(variables)):
            data.append({variables_str[loop]: variables[loop]})
            #print(data)
            pickle.dump(data,open(data_pickle,"wb"))
            #myworkspacetoo = pickle.load(open(data_pickle, "rb"))
            #print(myworkspacetoo)
            
            